### PR TITLE
Slice AnimationClip

### PR DESCRIPTION
**Pull request by Fyrestar**
To create a new clip from a existing from-to frame with optional new name. Some tools only export animations as a single clip.

### CODE REVIEW

File: src/animation/AnimationClip.js [hunk 1]
```diff
@@ -420,6 +420,45 @@ Object.assign( AnimationClip.prototype, {
...
+
+            var frames = to - from, duration = 0;
+
+            var tracks = [];
+
+            for ( var i = 0, l = this.tracks.length; i < l; i ++ ) {
+
+                var track = this.tracks[i];
+
+                var stride = track.values.length / track.times.length;
```
Comment (donmccurdy) [inline src/animation/AnimationClip.js]: Would use `track.getValueSize()` here.

File: src/animation/AnimationClip.js [hunk 2]
```diff
@@ -420,6 +420,45 @@ Object.assign( AnimationClip.prototype, {
...
+                    for ( var k = 0; k < stride; k ++ )
+                        values[t * stride + k] = track.values[ ( from + t ) * stride + k ];
+
+                }
+
+                duration = Math.max(times[ to ], duration);
+
+                tracks.push(
+                    new track.constructor( track.name, times, values, track.createInterpolant )
+                );
```
Comment (donmccurdy) [inline src/animation/AnimationClip.js]: It's possible that some of the tracks won't have any keyframes in the sliced range, so it might be best to keep empty tracks out of the new clip.

File: src/animation/AnimationClip.js [hunk 3]
```diff
@@ -420,6 +420,45 @@ Object.assign( AnimationClip.prototype, {
...
+                for ( var t = 0; t < frames; t ++ ) {
+
+                    times[t] = track.times[ from + t ] - start;
+
+                    for ( var k = 0; k < stride; k ++ )
+                        values[t * stride + k] = track.values[ ( from + t ) * stride + k ];
+
+                }
+
+                duration = Math.max(times[ to ], duration);
```
Comment (donmccurdy) [inline src/animation/AnimationClip.js]: Minor: Duration is computed automatically, so it might be simpler just to pass it `undefined` in the constructor.

**Review by donmccurdy — COMMENTED** [review]
Related: #13430

I'm OK with merging either of the PRs; mine would need a rebase anyway. Do you have an example model with multiple subclips, to test that this works as expected and gives smoothly-looping clips? There are some edge cases to be aware of — like:

```
-------------clip1-------------↓-------------clip2-------------↓-------------clip3-------------|
track1 ----|----|----|----|----|----|----|----|----|----|----|----|----|----|----|----|----|---|
track2----|------------------------------------------------------------------------------------|
track3----|----|----|----|----|-------------------------------------|----|----|----|----|----|-|
```

... where track2 and track3 actually do have values in clip2, but no keyframes. For a correctly authored animation that wouldn't be a problem, but automatic optimizations at export or import can mess things up...

File: src/animation/AnimationClip.js [hunk 4]
```diff
@@ -420,6 +420,45 @@ Object.assign( AnimationClip.prototype, {
 		return this;
 
 	},
+	
+	slice: function(from, to, name) {
+
+            var frames = to - from, duration = 0;
```
Comment (takahirox) [inline src/animation/AnimationClip.js]: Code style: We use hard tabs, not soft tabs.

File: src/animation/AnimationClip.js [hunk 5]
```diff
@@ -420,6 +420,45 @@ Object.assign( AnimationClip.prototype, {
 		return this;
 
 	},
+	
+	slice: function(from, to, name) {
```
Comment (takahirox) [inline src/animation/AnimationClip.js]: Code style: Extra spaces are necessary after opening bracket and before closing bracket `slice: function( from, to, name ) {`

https://github.com/mrdoob/three.js/wiki/Mr.doob's-Code-Style%E2%84%A2#blocks

File: src/animation/AnimationClip.js [hunk 6]
```diff
@@ -420,6 +420,45 @@ Object.assign( AnimationClip.prototype, {
...
+
+            var frames = to - from, duration = 0;
+
+            var tracks = [];
+
+            for ( var i = 0, l = this.tracks.length; i < l; i ++ ) {
+
+                var track = this.tracks[i];
```
Comment (takahirox) [inline src/animation/AnimationClip.js]: Code style: Extra spaces are necessary after opening bracket and before closing bracket.

https://github.com/mrdoob/three.js/wiki/Mr.doob's-Code-Style%E2%84%A2#arrays

**Comment by Fyrestar** [comment]
When i export animations with optimizations it will basically remove the start/end keyframe of the desired sequence of a track when smoothing it out. I'll change it to compute the values at these times,

**Comment by Mugen87** [comment]
Thanks for the work here but I think it's better to focus on #13430 (which is already tested and approved).
